Rubberduck.Inspections/Concrete/ShadowedDeclarationInspection.cs

```diff
@@ -0,0 +1,232 @@
+
+            var allUserProjects = UserDeclarations.OfType(DeclarationType.Project).Cast<ProjectDeclaration>();
+
+            foreach (var userProject in allUserProjects)
+            {
+                var referencedProjectIds = userProject.ProjectReferences.Select(reference => reference.ReferencedProjectId).ToHashSet();
+
+                var userDeclarations = UserDeclarations.Where(d => d.ProjectId == userProject.ProjectId);
+
+                // User has no control over build-in event handlers or their parameters, so we skip them
+                userDeclarations = userDeclarations.Where(d => !DeclarationIsPartOfBuiltInEventHandler(d, builtInEventHandlers));
+
+                foreach (var declaration in userDeclarations)
+                {
+                    var shadowedDeclaration = State.AllDeclarations.FirstOrDefault(d =>
+                        referencedProjectIds.Contains(d.ProjectId) && d.IdentifierName == declaration.IdentifierName && DeclarationCanBeShadowed(d, declaration));
+
+                    if (shadowedDeclaration != null)
+                    {
+                        issues.Add(new DeclarationInspectionResult(this,
+                            string.Format(InspectionsUI.ShadowedDeclarationInspectionResultFormat,
+                                RubberduckUI.ResourceManager.GetString("DeclarationType_" + declaration.DeclarationType, CultureInfo.CurrentUICulture),
+                                declaration.IdentifierName,
+                                RubberduckUI.ResourceManager.GetString("DeclarationType_" + shadowedDeclaration.DeclarationType,CultureInfo.CurrentUICulture),
+                                shadowedDeclaration.IdentifierName),
+                            declaration));
+                    }
+                } 
+            }
+
+            return issues;
+        }
+
+        private static bool DeclarationIsPartOfBuiltInEventHandler(Declaration declaration, ICollection<Declaration> builtInEventHandlers)
+        {
+            if (builtInEventHandlers.Contains(declaration))
+            {
+                return true;
+            }
+
+            var parameterDeclaration = declaration as ParameterDeclaration;
+
+            return parameterDeclaration != null && builtInEventHandlers.Contains(parameterDeclaration.ParentDeclaration);
+        }
+
+        private bool DeclarationCanBeShadowed(Declaration originalDeclaration, Declaration userDeclaration)
+        {
+            var originalDeclarationComponentType = originalDeclaration.QualifiedName.QualifiedModuleName.ComponentType;
+            var userDeclarationComponentType = userDeclaration.QualifiedName.QualifiedModuleName.ComponentType;
+
+            // It is not possible to directly access a Parameter, UDT Member or Label declared in another project
+            if (originalDeclaration.DeclarationType == DeclarationType.Parameter || originalDeclaration.DeclarationType == DeclarationType.UserDefinedTypeMember || 
```

Comment:
this seems to weed out a vast number of declarations. Shouldn't this filter be applied in the query across `State.AllDeclarations`?